Spark multilevel metrics collection refactoring.

cmnbroad opened the pull request:
- The main change is the use of a reducible multi level collector; currently there is only one consumer which is InsertizeMetricsCollector.
- There is a long comment and diagram at the top of MetricsCollectorSpark that describes the roles and relationships of the various component classes that make up a collector. Its a good place to start to understand these changes.
- The Spark and walker versions of CollectInsertSizeMetrics now use the same MultiLevelCollectorReducible-derived implementation (InsertSizeCollector), and exactly the same tests and expected results.
- All of the CollectInsertSizeMetricsSpark code and corresponding unit tests from the previous implementation are removed in this PR.
- Fixes https://github.com/broadinstitute/gatk/issues/1512.
- Both CollectMultipleMetrics (Spark and walker) implementations include support and tests for external/custom collectors.

droazen commented:
@cmnbroad Can you nominate a reviewer for this branch?

cmnbroad commented:
Nominating @akiezun. @lbergelson knows the metrics code pretty well too.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsArgumentCollection.java:
```diff
@@ -16,24 +16,36 @@
+    public String histogramPlotFile;
 
-    @Argument(doc = "Generate mean, sd and plots by trimming the data down to MEDIAN + maxMADTolerance*MEDIAN_ABSOLUTE_DEVIATION. " +
+    @Argument(doc="Generate mean, sd and plots by trimming the data down to MEDIAN + DEVIATIONS*MEDIAN_ABSOLUTE_DEVIATION. " +
```
why this change? the comment is now harder to understand, ie what is `DEVIATIONS` ?

cmnbroad replied:
Also a result of consolidating two similar but not identical implementations. I'll revert this to the previous doc string so it matches the argument name.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -1,50 +1,66 @@
-        extends MultiLevelCollector<InsertSizeMetrics, Integer, InsertSizeMetricsCollectorArgs>
-        implements Serializable {
-
+public class InsertSizeMetricsCollector
```
why not final anymore? Is it designed to be subclassed? there are no subclasses now and let's assume the will be none

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -1,50 +1,66 @@
+            final InsertSizeMetricsArgumentCollection inputArgs,
+            final SAMFileHeader samHeader)
+    {
+        super();
```
explicit `super()` calls in constructors are not required

akiezun replied:
can any of the arguments be `null`? If not, add checks. If yes, document it.

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -1,50 +1,66 @@
+     * @param samHeader SAMFileHeader for the input file
+     * @return ReadFilter to be used to filter records
+     */
+    public ReadFilter getReadFilter(SAMFileHeader samHeader) {
```
add null check for the header

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -1,50 +1,66 @@
+    // the record collectors. This method is called once per samRecord
     @Override
     protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceSequence refSeq) {
         final int insertSize = Math.abs(samRecord.getInferredInsertSize());
```
add a comment why the abs call is needed

cmnbroad replied:
I'm not sure why its this way - this code is in the existing walker implementation. The Picard code also does the same thing.

akiezun replied:
the reason is that the number is negative if the mate maps to lower position than the read. Would be useful to see this as a note here

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
-    protected PerUnitMetricCollector<InsertSizeMetrics, Integer, InsertSizeMetricsCollectorArgs> makeChildCollector(final String sample, final String library, final String readGroup) {
-        return new PerUnitInsertSizeMetricsCollector(sample, library, readGroup);
+    protected PerUnitInsertSizeMetricsCollector makeChildCollector(
+            final String sample,
```
check all arguments for null

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
-            return;
-        }
-
         super.acceptRecord(record, refSeq);
```
if all this method does is call super, then it's useless and we could delete it from here.

cmnbroad replied:
True that. Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
+     * @param source source of metrics to be merged into target
+     * @return single object representing the combined source and target objects
+     */
+    public InsertSizeMetricsCollector combine(InsertSizeMetricsCollector target, InsertSizeMetricsCollector source) {
```
check null for arguments

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
+    public PerUnitInsertSizeMetricsCollector combineUnit(
+            PerUnitInsertSizeMetricsCollector collector1,
+            PerUnitInsertSizeMetricsCollector collector2) {
+        return collector1.combine(collector2);
```
null check for args

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
+            final String inputName,
+            final AuthHolder authHolder)
+    {
+        super.finish();
```
this is a strange call. Usually, methods only call super if they want to invoke the method that is overriden by the current method. Here, the call is to a **different** method hierarchy. Can this simply become a virtual call to `finish()` rather than a non-virtual hardwired call to `MultiLevelCollectorReducible.finish()` ?

cmnbroad replied:
Yes. Done.

akiezun commented on src/test/resources/org/broadinstitute/hellbender/tools/picard/analysis/CollectInsertSizeMetrics/expectedQualityYieldOnInsertSizeMetrics.txt:
```diff
@@ -0,0 +1,10 @@
+## htsjdk.samtools.metrics.StringHeader
+# org.broadinstitute.hellbender.tools.spark.pipelines.metrics.CollectMultipleMetricsSpark  --output /var/folders/3g/0dh8mspj3pz8qpmm4qqqzblr0000gn/T/collectMultiMetricsTest381676565295870422/collectMultiMetrics --collectors CollectInsertSizeMetrics --collectors CollectQualityYieldMetrics --input /Users/cmn/projects/hellbender/src/test/resources/org/broadinstitute/hellbender/tools/picard/analysis/ ...
```
this file was clearly generated using hellbender test suite. is it the same as a file generated using picard on the same inputs? I think we should (whenever possible) use expected results that were created from picard (or gatk3 when appropriate)

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/MetricsCollectorSpark.java:
```diff
@@ -14,41 +14,72 @@
  *     collector.initialize(args);
  *
- *     ReadFilter filter == collector.getReadFilter();
+ *     ReadFilter filter == collector.getReadFilter(samFileHeader);
```
is this `samFileHeader` different from `getHeaderForReads()` 2 lines below? I think it's the same

cmnbroad replied:
Yes good catch, thats misleading. Fixed.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/MetricsCollectorSpark.java:
```diff
@@ -99,21 +131,16 @@ default ReadFilter getReadFilter(SAMFileHeader samHeader) {
      * gives the collector to serialize the resulting metrics, usually
      * to a metrics file.
+     * @param inputBaseName base name of the input file
+     * @param authHolder
```
can `authHolder` be null? If yes, document it.

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetrics.java:
```diff
@@ -71,4 +69,12 @@
     public int WIDTH_OF_90_PERCENT;
     /** The "width" of the bins, centered around the median, that encompass 100% of all read pairs. */
     public int WIDTH_OF_99_PERCENT;
+
```
Why is this not a javadoc?

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+
+    // kryo can't serialize an EnumMap so use a LinkedHashMap (also so we can maintain order
+    // the insertion order of entries since the combining code below counts on it
+    final Map<SamPairUtil.PairOrientation, Histogram<Integer>> Histograms = new LinkedHashMap<>();
```
lowercase names for fields

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+
+    //Explicitly sets the Histogram width, overriding automatic truncation of Histogram tail.
+    //Also, when calculating mean and stdev, only bins <= HISTOGRAM_WIDTH will be included.
+    private Integer HistogramWidth;
```
lower case. also, why Integer and not int? If null is significant, mention why (or change to `OptionalInt` to make it super explicit

cmnbroad replied:
This value comes from an optional command line argument that may be specified by the user. Added a comment.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+            final double deviations,
+            final Integer HistogramWidth) {
+
+        super();
```
remove all those calls to super() in constructors

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+            prefix = "All_Reads.";
+        }
+
+        Histograms.put(SamPairUtil.PairOrientation.FR,     new Histogram<Integer>("insert_size", prefix + "fr_count"));
```
use diamonds, ie `new Histogram<>`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public void addMetricsToFile(final MetricsFile<InsertSizeMetrics, Integer> file) {
```
add an `@Override` to make clear that it implements a method from interface (there's no `@Implement` so `@Override` has to do

cmnbroad replied:
Done, and removed unnecessary warning suppression.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+        Histograms.get(args.getPairOrientation()).increment(args.getInsertSize());
+    }
+
+    public void finish() { }
```
`@Override`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+        Histograms.put(SamPairUtil.PairOrientation.RF,     new Histogram<Integer>("insert_size", prefix + "rf_count"));
+    }
+
+    public void acceptRecord(final InsertSizeMetricsCollectorArgs args) {
```
`@Override`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    // Safely determine if two strings are equal, even if one or both are null.
+    private static boolean safeEqualStrings(final String source, final String target) {
```
delete this and use `Objects.equals`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+    // the insertion order of entries since the combining code below counts on it
+    final Map<SamPairUtil.PairOrientation, Histogram<Integer>> Histograms = new LinkedHashMap<>();
+
+    final String sample;
```
why are those fields no private? should be private whenever possible

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -0,0 +1,226 @@
+        this.HistogramWidth = HistogramWidth;
+        this.deviations = deviations;
+
+        String prefix = null;
```
exctract all this to a method to create a prefix

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
+
+        final RScriptExecutor executor = new RScriptExecutor();
+        executor.addScript(new Resource(R_SCRIPT,
+                org.broadinstitute.hellbender.tools.picard.analysis.CollectInsertSizeMetrics.class));
```
why use full name of this class?

cmnbroad replied:
Good point - I moved the R file to a metrics resource folder, and no longer need the qualified name.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollector.java:
```diff
@@ -53,135 +69,154 @@ protected InsertSizeMetricsCollectorArgs makeArg(SAMRecord samRecord, ReferenceS
+                histFile.getAbsolutePath(), // PDF graphics file
+                inputName                   // input bam file
+        );
+        if (inputArgs.histogramWidth != null)
```
braces here and everywhere

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/InsertSizeMetricsCollectorArgs.java:
```diff
@@ -2,13 +2,17 @@
-final class InsertSizeMetricsCollectorArgs {
+final class InsertSizeMetricsCollectorArgs implements Serializable {
+
+    public static final long serialVersionUID = 1L;
```
`serialVersionUID` should be `private static final long`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MetricsArgumentCollection.java:
```diff
@@ -1,19 +1,20 @@
  */
-public class MetricsArgumentCollection {
+public class MetricsArgumentCollection implements Serializable {
+    public static final long serialVersionUID = 1L;
```
`serialVersionUID` should be `private static final long`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+ *                 and PerUnitMetricCollector)
+ * @param <UNIT_COLLECTOR> Class to represent metrics collected at a single level
+ */
+public abstract class MultiLevelCollectorReducible<
```
why this name and not `MultiLevelReducibleCollector` ? Classes typically have names that are nouns, adjectives are sometimes used for interfaces like `Serializable`

cmnbroad replied:
Only to emphasize that its the same as MultiLevelCollector, except its reducible. But someday MultiLevelCollector will be gone so I'll rename it.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+
+    public static final String UNKNOWN = "unknown";
+
+    // A singleton UNIT_COLLECTOR that will accept all records
```
Make this a javadoc - here and elsewhere,  all such comments should be in javadoc style

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+    // A singleton UNIT_COLLECTOR that will accept all records
+    // TODO: This should be moved into the AllReadDistributor, since it represents the lone
+    // unit collector for that level, but its here since its currently accessed from outside
+    protected UNIT_COLLECTOR allReadCollector;
```
does this field need to be exposed like this and non encapsulated?

cmnbroad replied:
Made private.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+    // A list of Distributor that is at most length 4, 1 for each (ALL_READS, SAMPLE, LIBRARY, READ_GROUP)
+    // accumulation levels these will be listed in the order in which their children would be added to a
+    // metric file
+    protected List<Distributor> outputOrderedDistributors;
```
exposing field like this is not as good as encapsulating them

cmnbroad replied:
Made private.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+    public void combine(MultiLevelCollectorReducible<METRIC_TYPE, HISTOGRAM_KEY, ARGTYPE, UNIT_COLLECTOR> source)
+    {
+        if ((allReadCollector == null) != (source.allReadCollector == null)) {
+            throw new IllegalArgumentException("MultiLevelCollectors must be the same size to be combined");
```
somehow this messase seems weird given that it checks nullness of allReadCollector

cmnbroad replied:
Yeah, that is weird. Reworded and unified with the tests further down.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+            for(final SAMReadGroupRecord rg : rgRecs) {
+                final String key = getKey(rg);
+                if(!collectors.containsKey(key)) {
+                    collectors.put(key, makeCollector(rg));
```
use `putIfAbsent` instead of the `if`

cmnbroad replied:
Done.

akiezun replied:
good point. thanks

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+        //Make a PerUnitMetricCollector for this given Distributor
+        protected abstract UNIT_COLLECTOR makeCollector(final SAMReadGroupRecord rg);
+
+        protected abstract UNIT_COLLECTOR makeUnknownCollector();
```
add a comment about what this method does

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+        public Distributor(final List<SAMReadGroupRecord> rgRecs) {
+            collectors = new LinkedHashMap<>();
+            for(final SAMReadGroupRecord rg : rgRecs) {
+                final String key = getKey(rg);
```
the getKey and makeCollector methods are called during object initialization, which may result in hard to detect bugs because methods typically assume that the object has been created and initialized

cmnbroad replied:
Done. Separated out initialization into a separate method.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+         * @param reMap map entry merge (reMap) function for combining distributor map entries. This
+         *              is usually the combineUnit method of MultiLevelCollectorReducible.
+         */
+        public void combine(Distributor source, BiFunction<UNIT_COLLECTOR, UNIT_COLLECTOR, UNIT_COLLECTOR> reMap) {
```
why use `combine` in some places and `merge` in others?

cmnbroad replied:
I think the only place merge is used is when referring to arguments for the actual merge method on Map. Having said that, some of the comments in other places are not inconsistent. Will unify.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+    /** A dummy Distributor to handle the ALL_READS accumulation level.  No distribution is required
+     * since there should only ever be one PerUnitMetricCollector for ALL_READS.
+     */
+    private class AllReadsDistributor extends Distributor implements Serializable {
```
these non static inner classes make me uncomfortable a bit. Can they be made static ?

cmnbroad replied:
Yes, done, at the expense of making the code look much more complicated because all the type parameters have to be conferred on all of the inner classes, and the need for an explicit reference to the outer class instance to access the overloaded methods. Its the right thing to do though for serialization. The distributor hierarchy could probably be simplified more, but not as part of this pr.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+            return null;
+        }
+
+        public void acceptRecord(final ARGTYPE args, final SAMReadGroupRecord rg) {
```
`@Override`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+
+        @Override
+        public void finish() {
+            allReadCollector.finish();
```
is it ok to just ignore the finish method in super class? if yes, add a comment to the super class that it's ok to override and add a comment here that superclass is ignored on purpose

cmnbroad replied:
Yes, the all reads collector is a special case. However as part of the change to static inner classes, I changed the all reads distributor to use the same collector list the other distributors use, so this method overload is no longer necessary at all.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+
+        private static final long serialVersionUID = 1L;
+
+        public AllReadsDistributor(final List<SAMReadGroupRecord> rgRecs) {
```
the rgRegs are ignored, should you remove them from the constructor's signature?

cmnbroad replied:
Yes. In fact since I separated out initialization from construction, they're removed from all of the constructors.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+        public AllReadsDistributor(final List<SAMReadGroupRecord> rgRecs) {
+            // NOTE: since we're passing an empty list to the super class constructor,
+            // no collector will actually be added to the collector map; instead this
+            // class assumes a hardcoded allReadsCollector in all of it's other methods
```
`it's` -> `its`

cmnbroad replied:
Comment is removed and no longer necessary as part of the change to static inner classes.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+     * Construct a argument of ARGTYPE using the given SAMRecord and ReferenceSequence then pass
+     * this value to all collectors that should include this record
+     */
+    public void acceptRecord(final SAMRecord record, final ReferenceSequence refSeq) {
```
`@Override`

cmnbroad replied:
This one isn't an override in the Distributor hierarchy; its the original in the MultiLevelCollector hierarchy.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java:
```diff
@@ -0,0 +1,428 @@
+
+    /** Get the PerUnitMetricCollector that collects reads for all levels */
+    public UNIT_COLLECTOR getAllReadsCollector() {
+        return allReadCollector;
```
should rename the field to `allReadsCollector` or rename this method to `getAllReadCollector` because they are out of synch now

cmnbroad replied:
Made them all use allReads.

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/QualityYieldMetrics.java:
```diff
@@ -140,4 +140,12 @@ public QualityYieldMetrics finish() {
         return this;
     }
+
+    // Return a disambiguating name suffix to be used by the multiple collectors to
```
make this into javadoc

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -21,6 +21,8 @@
 import java.util.*;
 
 public final class AlignmentSummaryMetricsCollector extends SAMRecordAndReferenceMultiLevelCollector<AlignmentSummaryMetrics,  Long> {
+    private static final long serialVersionUID = 1L;
```
should the class be `Serializable`? right now it has a field `serialVersionUID` but it's not Serializable so the field is useless

cmnbroad replied:
No, its not necessary until its ported to the new framework. Removed.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/AlignmentSummaryMetricsCollector.java:
```diff
@@ -124,6 +126,8 @@ private boolean isAdapterSequence(final byte[] read) {
     }
 
     private class GroupAlignmentSummaryMetricsPerUnitMetricCollector implements PerUnitMetricCollector<AlignmentSummaryMetrics, Long, SAMRecordAndReference> {
```
can this class be static?

cmnbroad replied:
Not without making a bunch of other changes. All of these (other Picard) metrics classes will all have to be ported to hellbender style and the new metrics framework, but those will be separate PRs.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/CollectAlignmentSummaryMetrics.java:
```diff
@@ -48,6 +49,10 @@
 
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "File to write the output to.")
+    public File OUTPUT;
```
fields have lowercase names

cmnbroad replied:
Agreed, but the only change here was to push the definition down from SinglePassSam into the individual classes as an interim step until these classes are all ported to the new framework. If we change it to be lower case, all of the references here and in all of the other classes it got pushed down to would also have to be changed. And this class also has 5 other fields all in upper case, and doesn't conform in other ways. I'd propose we wait and make those changes in the PRs where these tools and classes are made GATK compliant, rather than adding all of those changes to this PR.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/CollectBaseDistributionByCycle.java:
```diff
@@ -26,6 +27,10 @@
 
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "File to write the output to.")
+    public File OUTPUT;
```
lowercase

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -40,6 +41,10 @@
 
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "File to write the output to.")
+    public File OUTPUT;
```
lowercase

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/MeanQualityByCycle.java:
```diff
@@ -36,6 +37,10 @@
 
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "File to write the output to.")
+    public File OUTPUT;
```
lowercase

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/QualityScoreDistribution.java:
```diff
@@ -33,6 +34,10 @@
 
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "File to write the output to.")
+    public File OUTPUT;
```
lowercase

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/artifacts/CollectSequencingArtifactMetrics.java:
```diff
@@ -51,6 +52,10 @@
 
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "File to write the output to.")
+    public File OUTPUT;
```
lowercase. 

btw, why was the output field moved from `SinglePassSamProgram` down to subclasses?

cmnbroad replied:
Because the collectors need be instantiable in multiple contexts, so all of their args need to be in their own argument collection, and available in all contexts. If I leave it in SinglePassSam, it will be a duplicate arg in that context. As the other collectors are moved to this pattern, the individual ones that were moved down will go away and there will only be one again (in the MetricsArgumentCollection base class).

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/picard/analysis/directed/CollectRnaSeqMetrics.java:
```diff
@@ -37,6 +38,10 @@
 
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "File to write the output to.")
+    public File OUTPUT;
```
lowercase

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectMultipleMetricsSpark.java:
```diff
@@ -93,68 +94,95 @@
+        List<SparkCollectorProvider> collectorsToRun = new ArrayList<>();
+        if (externalCollectors == null) {
+            if (userCollectors.size() == 0) { // run them all
+                collectorsToRun = Arrays.asList(SparkCollectors.values());
```
can you just `return Arrays.asList(SparkCollectors.values())`

cmnbroad replied:
Done.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/MetricsCollectorSparkTool.java:
```diff
@@ -20,7 +20,7 @@
  * - Declare an instance of a collector that implements the MetricsCollectorSpark interface
  * - Declare an instance of the input argument collection (of type T) for the collector
- * - Implement or override the getReadFilter, initialize, collectMetrics and finishCollection
+ * - Implement or override the getReadFilter, initialize, collectMetrics and saveMetrics
```
all those should be javadoc links like this `{@link #method}`

cmnbroad replied:
Done.

akiezun commented on src/test/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducibleUnitTest.java:
```diff
@@ -0,0 +1,254 @@
+    public static File TESTFILE = new File("src/test/resources/org/broadinstitute/hellbender/metrics/test.sam");
+
+    public String noneOrStr(final String str) {
+        final String out;
```
`return str == null ? "" : str;`

cmnbroad replied:
Done.

akiezun commented on src/test/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducibleUnitTest.java:
```diff
@@ -0,0 +1,254 @@
+    /** We will just Tally up the number of times records were added to this metric and change FINISHED
+     * to true when FINISHED is called
+     */
+    class TotalNumberMetric extends MultiLevelMetrics {
```
`static final class`?

cmnbroad replied:
Done.

akiezun commented on src/test/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducibleUnitTest.java:
```diff
@@ -0,0 +1,254 @@
+        //distribute the reads across the two collectors
+        int count = 1;
+        for (final SAMRecord rec : in) {
+            if (count % 2 == 0) {
```
how is this supposed to work? `count` is never incremented

cmnbroad replied:
Fixed.

akiezun commented:
this is a whopper. @lbergelson can you have a look at it too?

droazen commented:
Since this new framework is pretty independent of much of the rest of the codebase, I'd recommend we merge once @akiezun's comments are addressed and @lbergelson gives the thumb's up, assuming no one has major issues with the overall design, instead of letting the review drag on too long (especially as @cmnbroad is away next week!)

davidbernick commented:
Can one of the admins verify this patch?

akiezun commented:
@droazen that sounds good to me

cmnbroad commented:
I addressed every comment except two: the suggested "putIfAbsent" change (see the comment there - I don't think we should do it), and I didn't change the case of the OUTPUT field that I moved down from SinglePassSAM to several Picard metrics classes (also see the note) since doing so would require changing all of the references in those classes, and they have many other fields that need similar changes. I'd propose that those changes be done in separate PRs, as those classes are ported to this framework.

cmnbroad commented:
Added a couple more javadoc updates. Not squashed, otherwise all comments have been addressed (except the two described above).

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/PerUnitInsertSizeMetricsCollector.java:
```diff
@@ -162,65 +151,63 @@ public void addMetricsToFile(final MetricsFile<InsertSizeMetrics, Integer> file)
-        validateEquals(this.library, sourceCollector.library, validationMessage);
-        validateEquals(this.readGroup, sourceCollector.readGroup, validationMessage);
+        Utils.nonNull(sourceCollector);
+        Objects.equals(this.sample, sourceCollector.sample);
```
hmm, this does not work. Objects.equals just returns false when the args are not equal but this should blow up. Need to wrap in an if statement

droazen replied:
Fixed

akiezun commented on src/main/java/org/broadinstitute/hellbender/metrics/QualityYieldMetricsArgumentCollection.java:
```diff
@@ -17,6 +17,6 @@
             doc = "If available in the OQ tag, use the original quality scores " +
                     "as inputs instead of the quality scores in the QUAL field.")
-    public boolean useOriginalQualities = false;
+    public boolean useOriginalQualities = true;
```
why this change to the default value?

droazen replied:
Reverted, we'll see if tests pass

droazen replied:
Tests fail -- taking a look at the failures now

droazen replied:
Turns out that in picard this defaults to true, which is likely why @cmnbroad changed it. Setting it back to true.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/metrics/MetricsArgumentCollection.java:
```diff
@@ -1,19 +1,20 @@
  * Base class for defining a set of metrics collector arguments. The members should be
  * instantiable as command line arguments.
  */
-public class MetricsArgumentCollection {
```
Should this be an abstract class?  Will anyone want to instantiate one of the directly?

akiezun commented:
I found 2 issues in the revised code: Object.equals should be wrapped in an if-notEquals-then-blowup and an unexplained change in default. 

Other than those two, 👍 from me

lbergelson commented:
I didn't do a line by line review since it seemed like Adam did a very thorough job there.  The overall structure looks good to me though.  It seems like metrics collection would be a good candidate for being made a plugin system, as in the ReadFilter branch.  That should definitely be a different pass on this though if necessary.

akiezun commented:
who's taking over the branch?

droazen commented:
@akiezun I'll take care of addressing your last few comments on this branch and merging it in.

droazen commented:
Tests pass -- merging
